### PR TITLE
[ CB ][ SD ] Support streaming with using `stop_strings` and `include_stop_strings`

### PR DESCRIPTION
**Pull request by iefode**
*Details:*:
* Implement streaming with using `stop_strings` in CB like pipelines
* Change `stop_string_match` logic to encode them only once per request
* Do not stream tokens which are matched to the part of a `stop_string` (Tests was a bit changes in this case according HF does not support exclude `stop_strings`)

*Tickets:*
* CVS-158463

### CODE REVIEW

File: src/cpp/src/continuous_batching_impl.cpp [hunk 1]
```diff
@@ -297,11 +296,7 @@ ContinuousBatchingPipeline::ContinuousBatchingImpl::generate(const std::vector<o
         streamer_ptr->end();
     }
 
-    if (!continue_generation) {
-        drop_requests();
```
Comment (ilya-lavrenov) [inline src/cpp/src/continuous_batching_impl.cpp:301]: why have you removed `drop_requests` here ? Looks like `continue_generation` should be assigned to `!streamer_ptr->put` return value?

Looks like the logic was broken in https://github.com/openvinotoolkit/openvino.genai/pull/1377

File: src/cpp/src/sampler.cpp [hunk 2]
```diff
@@ -85,75 +85,65 @@ std::string clean_wrapped_text(const std::string& wrapped_text, const std::strin
     return clean_text;
 }
 
+std::vector<int64_t> encode_and_process_string(const std::string& stop_string, ov::genai::Tokenizer& tokenizer) {
+    // encode stop_string
+    ov::Tensor ov_encoded_stop_string = tokenizer.encode(stop_string).input_ids;
+    size_t tensor_size = ov_encoded_stop_string.get_size();
+    std::vector<int64_t> source_encoded_stop_string(tensor_size), encoded_stop_string;
+    std::copy_n(ov_encoded_stop_string.data<int64_t>(), tensor_size, source_encoded_stop_string.begin());
+    // remove special symbols
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:94]: should we encode with add_special_tokens=false to avoid this post-processing ?

Comment (iefode) [reply src/cpp/src/sampler.cpp:94]: Did not get your point. We should remove all special symbols from `stop_strings`

**Comment by ilya-lavrenov** [comment]
Please, add tests on streaming scenario, which you have fixed.

File: samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp [hunk 3]
```diff
@@ -33,7 +33,8 @@ int main(int argc, char* argv[]) try {
         main_model_path,
         main_device,
         ov::genai::draft_model(draft_model_path, draft_device),
-        ov::genai::scheduler_config(scheduler_config));
+        ov::genai::scheduler_config(scheduler_config)
+    );
```
Comment (ilya-lavrenov) [inline samples/cpp/speculative_decoding_lm/speculative_decoding_lm.cpp:37]: please, revert this change as it creates merge conflicts with https://github.com/openvinotoolkit/openvino.genai/pull/1364 where this lone is removed.

File: samples/cpp/text_generation/greedy_causal_lm.cpp [hunk 4]
```diff
@@ -13,7 +13,7 @@ int main(int argc, char* argv[]) try {
 
     ov::genai::LLMPipeline pipe(models_path, device);
     ov::genai::GenerationConfig config;
-    config.max_new_tokens = 100;
+    config.max_new_tokens = 100;
```
Comment (ilya-lavrenov) [inline samples/cpp/text_generation/greedy_causal_lm.cpp:16]: please, revert

File: src/cpp/src/sampler.hpp [hunk 5]
```diff
@@ -57,6 +57,8 @@ class Sampler {
     std::mt19937 rng_engine;
     // { request_id, logit_processor }
     std::map<uint64_t, LogitProcessor> m_logit_processors;
+    // { request_id, { max_encoded_len, { stop_strings }}}
+    std::map<int64_t, std::pair<size_t, std::set<std::string>>> m_stop_strings;
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.hpp:61]: you need to clean this information in `clear_request_info` once request has finished / dropped / etc

File: src/cpp/src/sequence_group.hpp [hunk 6]
```diff
@@ -126,23 +126,31 @@ class Sequence {
...
 
-        auto generated_token_id = get_generated_ids();
-        auto generated_log_probs = get_generated_log_probs();
+            auto generated_token_id = get_generated_ids();
+            auto generated_log_probs = get_generated_log_probs();
 
-        OPENVINO_ASSERT(get_generated_len() >= token_cnt);
-        auto offset = get_generated_len() - token_cnt;
+            if (get_generated_len() < token_cnt) {
+                auto a = 0;
```
Comment (ilya-lavrenov) [inline src/cpp/src/sequence_group.hpp:139]: remove this code?

File: src/cpp/src/sampler.cpp [hunk 7]
```diff
@@ -392,19 +382,19 @@ void Sampler::GroupBeamSearcher::select_next_tokens(const ov::Tensor& logits, Sa
...
                         continue;
 
-                    if(!m_parameters.include_stop_str_in_output) {
-                        // remove tokens that match stop_string from output (last token is not included in candidate.m_sequence at this point)
-                        candidate.m_sequence->remove_last_tokens(num_last_matched_tokens - 1);
-                    }
+                    // remove tokens that match stop_string from output (last token is not included in candidate.m_sequence at this point)
+                    candidate.m_sequence->remove_last_tokens(match_result.to_remove);
+
+                    // candidate.m_token_id = match_result.last_token_id;
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:394]: remove commented code?

File: src/cpp/src/sampler.cpp [hunk 8]
```diff
@@ -576,10 +566,13 @@ std::vector<int64_t> Sampler::_try_finish_generation(SequenceGroup::Ptr & sequen
...
 
         if (!sampling_params.stop_strings.empty()) {
-            int num_matched_last_tokens = match_stop_string(m_tokenizer, running_sequence->get_generated_ids(), sampling_params.stop_strings);
-            if (num_matched_last_tokens) {
-                if (!sampling_params.include_stop_str_in_output)
-                    running_sequence->remove_last_tokens(num_matched_last_tokens);
+            auto& stop_strings = m_stop_strings.at(sequence_group->get_request_id());
+            auto match_result = match_stop_string(m_tokenizer, running_sequence->get_generated_ids(), stop_strings, sampling_params.include_stop_str_in_output);
+            if (match_result.is_matched) {
+                if (match_result.to_remove > 0) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:572]: we don't need to check for `> 0` and in case of `== 0` method `remove_last_tokens` also works correctly

**Review by ilya-lavrenov — APPROVED** [review]
